File: src/main/java/pl/yalgrin/playnite/simplesync/service/ChangeListenerService.java
package pl.yalgrin.playnite.simplesync.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import pl.yalgrin.playnite.simplesync.dto.ChangeDTO;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.publisher.Sinks;
import reactor.core.scheduler.Scheduler;
import reactor.core.scheduler.Schedulers;

import java.util.List;

@Service
@Slf4j
public class ChangeListenerService {
    private final Sinks.Many<ChangeDTO> sink = Sinks.many().multicast().directBestEffort();
    private final Scheduler scheduler = Schedulers.newSingle("change-emitter");

    public Mono<Void> publishChange(ChangeDTO dto) {
        return Mono.fromRunnable(() -> {
            log.debug("publishing {}", dto);
            sink.tryEmitNext(dto);
        }).subscribeOn(scheduler).then();
    }

    public Mono<Void> publishChanges(List<ChangeDTO> dtoList) {
        return Flux.fromIterable(dtoList).concatMap(this::publishChange).then();
    }

    public Flux<ChangeDTO> modificationFlux() {
        return sink.asFlux();
    }
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/change/service/ChangeListenerService.kt
package pl.yalgrin.playnite.simplesync.change.service

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import pl.yalgrin.playnite.simplesync.dto.ChangeDTO
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.publisher.Sinks
import reactor.core.scheduler.Schedulers
import reactor.kotlin.core.publisher.toFlux

@Service
class ChangeListenerService {
    private val sink = Sinks.many().multicast().directBestEffort<ChangeDTO>()
    private val scheduler = Schedulers.newSingle("change-emitter")

    companion object {
        private val log = LoggerFactory.getLogger(ChangeListenerService::class.java)
    }

    fun publishChange(dto: ChangeDTO): Mono<Void> {
        return Mono.fromRunnable<Unit> {
            log.debug("publishing {}", dto)
            sink.tryEmitNext(dto)
        }.subscribeOn(scheduler).then()
    }

    fun publishChanges(dtoList: List<ChangeDTO>): Mono<Void> {
        return dtoList.toFlux()
            .concatMap { dto -> publishChange(dto) }
            .then()
    }

    fun modificationFlux(): Flux<ChangeDTO> {
        return sink.asFlux()
    }
}